### PR TITLE
Write rules to construct headers for jpeg-turbo using existing templates.

### LINKED ISSUE #6442: Build on windows using bazel broken
Operating System: Windows
bazel 0.4.2, issue is seen at master branch.

It looks like sometime after our update to libjpeg turbo, our build has been broken in windows.
I am looking into it, but I cannot see a clear resolution to the problem.
The error message says:
http://ci.tensorflow.org/job/tf-master-win-bzl/218/console
```
00:04:27.376 c:\tmp\_bazel_system\rrc05caq\execroot\tf-master-win-bzl\external\jpeg\jmorecfg.h(242): error C2371: 'boolean': redefinition; different basic types
00:04:27.376 c:\program files (x86)\windows kits\10\include\10.0.14393.0\shared\rpcndr.h(193): note: see declaration of 'boolean'
```

I checked jmorecfg.sh, and tried to use `#define HAVE_BOOLEAN` to get around the issue. However that causes:
http://ci.tensorflow.org/view/Experimental/job/exp-win-bzl/4/console

Creating the issue to track a quick resolution to the problem.

CC Our in house windows experts:
@meteorcloudy @mrry 

@guschmue @vit-stepanovs Do you have any ideas how we can solve this?
We would like to also upgrade cmake build to use libjpeg-turbo, but before that I want to make sure it builds when we do.

### PR DESCRIPTION
**Pull request by gunan**
Fixes #6442 
Test build:
http://ci.tensorflow.org/view/Experimental/job/exp-win-bzl/8/console

Had to use a few hacks to get this fully working.
Bazel did not allow me to add a `dict` type with a selector.
It even did not let me use a selector in `expand_template`'s substitutions field.
So I expanded the selector in genrule to accomodate simd/nosimd differences.

I think I can also get rid of my `expand_template.bzl` , and use `expand_header_template` in `llvm.bzl`.
But I am not sure how nice cross 3p usage. Maybe I can create a folder `third_party/skylark` and pull the skylark rules we are defining to that folder. Or simply `third_party/common.bzl`
Please let me know which looks nicer.

### CODE REVIEW

**Comment by gunan** [comment]
All regular tests seem to have passed.
In windows, build has succeeded, but due to prolonged red build, some test failures are there.

File: third_party/jpeg/expand_template.bzl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  )
+
+Args:
+  name: The name of the rule.
+  template: The template file to expand
+  out: The destination of the expanded file
+  substitutions: A dictionary mapping strings to their substitutions
+"""
+
+def expand_template_impl(ctx):
```
Comment (jhseu) [inline third_party/jpeg/expand_template.bzl]: Can we put this file in //third_party? I actually need it for something too.

**Comment by gunan** [comment]
Done, now the expand template macro is under third_party:common.bzl
PTAL.

File: third_party/llvm/llvm.bzl [hunk 2]
```diff
@@ -46,48 +46,6 @@ def gentbl(name, tblgen, td_file, td_srcs, tbl_outs, library = True, **kwargs):
...
-#   substitutions: A dictionary mapping strings to their substitutions
-
-def expand_header_template_impl(ctx):
-  ctx.template_action(
-      template = ctx.file.template,
-      output = ctx.outputs.out,
-      substitutions = ctx.attr.substitutions,
-  )
-
-expand_header_template = rule(
```
Comment (jart) [inline third_party/llvm/llvm.bzl]: Don't you need a load() statement at the top of this file now?

File: third_party/common.bzl [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#   substitutions: A dictionary mapping strings to their substitutions
+
+def expand_header_template_impl(ctx):
+  ctx.template_action(
+      template = ctx.file.template,
+      output = ctx.outputs.out,
+      substitutions = ctx.attr.substitutions,
+  )
+
+expand_header_template = rule(
```
Comment (jart) [inline third_party/common.bzl]: Any chance we can rename this `template_rule`?

**Review by jart — CHANGES_REQUESTED** [review]

File: third_party/jpeg/jpeg.BUILD [hunk 4]
```diff
@@ -5,6 +5,8 @@ licenses(["notice"])  # custom notice-style license, see LICENSE.md
 
 exports_files(["LICENSE.md"])
 
+load("@//third_party:common.bzl", "expand_header_template")
```
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: I guess on that one, we do not really have an alternative option?
I think one thing we can do is to define these in either tensorflow/tensorflow.bzl, or move this file to tensorflow/common.bzl?

Comment (jart) [inline third_party/jpeg/jpeg.BUILD]: It's currently not possible for other Bazel projects to depend on TensorFlow for other reasons, such as our ./configure being mandatory. So the way I see it, we can just add issue 1248 to our list of blockers. We can fix this later.

File: third_party/jpeg/jpeg.BUILD [hunk 5]
```diff
@@ -274,50 +276,118 @@ cc_library(
...
-          "#else\n" +
-          "#define SIZEOF_SIZE_T 4\n" +
-          "#endif\n" +
-          "EOF",
+    srcs = [
+        ":jconfigint_win",
+        ":jconfigint_nowin",
+    ],
+    cmd = select({
+        ":windows": "cp $(locations :jconfigint_win) $(@D)/jconfigint.h",
```
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Done.
One more thing, it turns out if I use "jconfigint_nowin.h" in srcs, I have to refer to it the same way here as I do in srcs.

File: third_party/common.bzl (hunk 3 above)
Comment (gunan) [inline third_party/common.bzl]: do you mean s/expand_header_template/template_rule/ ?

File: third_party/jpeg/jpeg.BUILD [hunk 6]
```diff
@@ -274,50 +276,118 @@ cc_library(
...
-          "#define PACKAGE_NAME \"libjpeg-turbo\"\n" +
-          "#define VERSION \"1.5.1\"\n" +
-          "#if (__WORDSIZE==64 && !defined(__native_client__)) || defined(_WIN64)\n" +
-          "#define SIZEOF_SIZE_T 8\n" +
-          "#else\n" +
-          "#define SIZEOF_SIZE_T 4\n" +
-          "#endif\n" +
-          "EOF",
+    srcs = [
+        ":jconfigint_win",
```
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Done

File: third_party/jpeg/jpeg.BUILD (hunk 5 above)
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Done

File: third_party/llvm/llvm.bzl (hunk 2 above)
Comment (gunan) [inline third_party/llvm/llvm.bzl]: not this specific file, but llvm.BUILD uses expand_header_template, so I added a load statement there.

File: third_party/common.bzl [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    implementation = expand_header_template_impl,
+    attrs = {
+        "template": attr.label(mandatory=True, allow_files=True, single_file=True),
```
Comment (gunan) [inline third_party/common.bzl]: Done

File: third_party/jpeg/jpeg.BUILD [hunk 8]
```diff
@@ -274,50 +276,118 @@ cc_library(
...
+    name = "jconfigint_nowin",
+    out = "jconfigint_nowin.h",
+    substitutions = {
+        "#undef BUILD" : "#define BUILD \"20161115\"",
+        "#undef inline" : "",
+        "#undef INLINE" : "#define INLINE inline __attribute__((always_inline))",
+        "#undef PACKAGE_NAME" : "#define PACKAGE_NAME \"libjpeg-turbo\"",
+        "#undef VERSION" : "#define VERSION \"1.5.1\"",
+        "#undef SIZEOF_SIZE_T" : "#if (__WORDSIZE==64 && !defined(__native_client__))\n" +
+            "#define SIZEOF_SIZE_T 8\n" +
```
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Done.
It modified the indentation which looks more consistent with python style.
So I am happy with what buildifier did.

File: third_party/jpeg/jpeg.BUILD [hunk 9]
```diff
@@ -274,50 +276,118 @@ cc_library(
...
+    out = "jconfigint_win.h",
+    substitutions = {
+        "@VERSION@" : "1.5.1",
+        "@BUILD@" : "20161115",
+        "@CMAKE_PROJECT_NAME@" : "libjpeg-turbo",
+    },
+    template = "win/jconfigint.h.in",
+)
+
+jconfig_nowin_common_substitutions = {
```
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Done.

File: third_party/jpeg/jpeg.BUILD (hunk 4 above)
Comment (gunan) [inline third_party/jpeg/jpeg.BUILD]: Acknowledged.
Do you want me to leave a comment here about that?

File: third_party/common.bzl (hunk 3 above)
Comment (gunan) [inline third_party/common.bzl]: Ping!
Wanted to bump this one.

File: third_party/jpeg/jpeg.BUILD (hunk 4 above)
Comment (jart) [inline third_party/jpeg/jpeg.BUILD]: Probably not worthwhile. We both know what needs to be done. We'll do it as soon as we can.

File: third_party/common.bzl (hunk 3 above)
Comment (jart) [inline third_party/common.bzl]: Yes.

**Review by jart — APPROVED** [review]

File: third_party/common.bzl (hunk 3 above)
Comment (gunan) [inline third_party/common.bzl]: Done

**Comment by gunan** [comment]
All changes done.
Pending test results to merge.

**Pull request merged**